src/daemon/socks5.rs

```diff
@@ -0,0 +1,121 @@
+
+use crate::{
+    config::{Fallback, Socks5},
+    socket::{Endpoint, Socket},
+    stream::Stream,
+};
+
+use super::DaemonContext;
+
+pub async fn socks5_loop(ctx: DaemonContext, socks5_cfg: Socks5) -> anyhow::Result<()> {
+    log::debug!("socks5 loop");
+    let tcp_listener = TcpListener::bind(SocketAddrV4::new(
+        "127.0.0.1".parse()?,
+        socks5_cfg.listen_port,
+    ))
+    .await?;
+    let fallback = socks5_cfg.fallback;
+    let reaper = TaskReaper::new();
+
+    loop {
+        let ctx = ctx.clone();
+        let fallback = fallback.clone();
+        let (tcp_stream, _) = tcp_listener.accept().await?;
+
+        reaper.attach(smol::spawn(async move {
+            tcp_stream.set_nodelay(true)?;
+            let _handshake = read_handshake(tcp_stream.clone()).await?;
+            write_auth_method(tcp_stream.clone(), SocksV5AuthMethod::Noauth).await?;
+            let request = read_request(tcp_stream.clone()).await?;
+            let port = request.port;
+            let domain: String = match &request.host {
+                SocksV5Host::Domain(dom) => String::from_utf8_lossy(dom).parse()?,
+                SocksV5Host::Ipv4(v4) => {
+                    let v4addr = Ipv4Addr::new(v4[0], v4[1], v4[2], v4[3]);
+                    v4addr.to_string()
+                }
+                _ => anyhow::bail!("IPv6 not supported"),
+            };
+            let addr = format!("{domain}:{port}");
+
+            write_request_status(
+                tcp_stream.clone(),
+                SocksV5RequestStatus::Success,
+                request.host,
+                port,
+            )
+            .await?;
+
+            log::info!("socks5 received request for {addr}");
+
+            let mut split_domain = domain.split('.');
+            let top_level = split_domain.clone().last();
+
+            if let Some(top) = top_level {
+                if top == "haven" {
+                    let endpoint = Endpoint::new(
+                        Fingerprint::from_str(
+                            split_domain.next().context("invalid Earendil address")?,
+                        )?,
+                        port.into(),
+                    );
+                    let earendil_skt =
+                        Socket::bind_haven_internal(ctx.clone(), ctx.identity, None, None);
+                    let earendil_stream = Stream::connect(earendil_skt, endpoint).await?;
+
+                    io::copy(tcp_stream.clone(), &mut earendil_stream.clone())
+                        .race(io::copy(earendil_stream.clone(), &mut tcp_stream.clone()))
+                        .await?;
+                } else {
+                    match fallback {
+                        Fallback::Block => return Ok(()),
+                        Fallback::PassThrough => {
+                            let mut addrs: Vec<SocketAddr> = addr.to_socket_addrs()?.collect();
+                            let passthrough_stream = TcpStream::connect(addrs.pop().context(
```

Comment:
I looked at the docs again, and you can actually pass a string `hostname:port` directly to TcpStream::connect, sidestepping all of this